src/radical/pilot/raptor/worker_default.py

```diff
@@ -0,0 +1,724 @@
+      # self._log.debug('gpus  %s', str(self._resources['gpus']))
+
+        # resources are initially all free
+        self._res_evt.set()
+
+      # # create a multiprocessing pool with `cpn` worker processors.  Set
+      # # `maxtasksperchild` to `1` so that we get a fresh process for each
+      # # task.  That will also allow us to run command lines via `exec`,
+      # # effectively replacing the worker process in the pool for a specific
+      # # task.
+      # #
+      # # We use a `fork` context to inherit log and profile handles.
+      # #
+      # # NOTE: The mp documentation is wrong; mp.Pool does *not* have a context
+      # #       parameters.  Instead, the Pool has to be created within
+      # #       a context.
+      # ctx = mp.get_context('fork')
+      # self._pool = ctx.Pool(processes=self._n_cores,
+      #                       initializer=None,
+      #                       maxtasksperchild=1)
+      # NOTE: a multiprocessing pool won't work, as pickle is not able to
+      #       serialize our worker object.  So we use our own process pool.
+      #       It's not much of a loss since we want to respawn new processes for
+      #       each task anyway (to improve isolation).
+        self._pool  = dict()  # map task uid to process instance
+        self._plock = ru.Lock('p' + self._uid)  # lock _pool
+
+        # We also create a queue for communicating results back, and a thread to
+        # watch that queue
+        self._result_queue  = mp.Queue()
+        self._result_thread = mt.Thread(target=self._result_watcher)
+        self._result_thread.daemon = True
+        self._result_thread.start()
+
+        # run worker initialization *before* starting to work on requests.
+        # the worker provides three builtin methods:
+        #     eval:  evaluate a piece of python code with `eval`
+        #     exec:  evaluate a piece of python code with `exec`
+        #     call:  execute  a method or function call
+        #     proc:  execute  a command line (fork/exec)
+        #     shell: execute  a shell command
+        self.register_mode('eval',  self._eval)
+        self.register_mode('exec',  self._exec)
+        self.register_mode('call',  self._call)
+        self.register_mode('proc',  self._proc)
+        self.register_mode('shell', self._shell)
+
+        self.pre_exec()
+
+        # prepare base env dict used for all tasks
+        self._task_env = dict()
+        for k,v in os.environ.items():
+            if k.startswith('RP_'):
+                self._task_env[k] = v
+
+
+    # --------------------------------------------------------------------------
+    #
+    def pre_exec(self):
+        '''
+        This method can be overloaded by the Worker implementation to run any
+        pre_exec commands before spawning worker processes.
+        '''
+
+        pass
+
+
+    # --------------------------------------------------------------------------
+    #
+    def register_mode(self, name, executor):
```

Comment:
Can you please document the signature of `executor` and point to more information about how the inputs to `executor` (`data`) are composed?